yt/yt/client/table_client/helpers.cpp

```diff
@@ -1568,4 +1569,49 @@ REGISTER_INTERMEDIATE_PROTO_INTEROP_BYTES_FIELD_REPRESENTATION(
 
 ////////////////////////////////////////////////////////////////////////////////
 
+TUnversionedValueRangeTruncationResult TruncateUnversionedValues(TUnversionedValueRange values, i64 size)
+{
+    auto rowBuffer = New<TRowBuffer>();
+
+    std::vector<TUnversionedValue> truncatedValues;
+    truncatedValues.reserve(values.size());
+
+    i64 resultSize = 0;
+    // Indicates the fact that we have truncated something.
+    // After this moment all next values have to be replaced by nulls to mantain sorting properties.
```

Comment:
That does not look good. Consider a table with extremely heavy first column and some light columns. After truncation, light columns will become null, so sorting by light columns will be extremely inefficient due to a lack of adequate samples.

Reply:
Discussed in person and agreed on the following qualities of the resulting algorithm:
* Primitive values should always be emitted into samples. 
* Short strings/composites should almost always be emitted into samples. 
* Long strings can be truncated.

Truncating each non-primitive value to (maxTotalSize - sizeOfPrimitiveValues) / (nonPrimitiveValueCount) should be fine (even though it might lead to some long values being excessively truncated).

In practice, key values of more than 16KiB will lead to the whole sort job failing.

We do not need to "clip" rows (replace all following values with nulls after a truncated value) when emitting samples, but we need to do need it when reading them (since we are actually reading the subset of columns which forms the comparison key).